### ISSUE TITLE
BUG: Newton's method + R space + Fieldsplit throws PyOP2 error

### ISSUE THREAD

**Issue by ABaierReinio** [firedrakeproject/firedrake#3837]
When using Newton's method on a form that involves R spaces, and doing a fieldsplit, a PyOP2 error is thrown.

MWE below:

```
from firedrake import *

M = UnitSquareMesh(5, 5)

V = FunctionSpace(M, "CG", 1)
R = FunctionSpace(M, "R", 0)
Z = V * R

func = Function(Z)
u, l = split(func)
v, w = TestFunctions(Z)

F = (
    (u + u**2 - 8) * v * dx
    + l * w * dx
)

newton_solver_parameters = {"snes_type" : "newtonls",
                            "snes_monitor" : "",
                            "snes_converged_reason" : "",
                            "mat_type" : "nest",
                            "ksp_type" : "fgmres",
                            "ksp_monitor" : "",
                            "pc_type": "fieldsplit",
                            "pc_fieldsplit_type": "schur",
                            "pc_fieldsplit_schur_fact_type": "full",
                            "pc_fieldsplit_0_fields": "0",
                            "pc_fieldsplit_1_fields" : "1",
                            "fieldsplit_0_ksp_type" : "preonly",
                            "fieldsplit_0_pc_type" : "lu",
                            "fieldsplit_0_pc_factor_mat_solver_type" : "mumps",
                            "fieldsplit_1_ksp_type" : "gmres",
                            "fieldsplit_1_max_it" : 1,
                            "fieldsplit_1_convergence_test" : "skip"
                            }

solve(F == 0, func, solver_parameters=newton_solver_parameters)
```

The script successfully does an iteration of Newton's method, but it then throws the error

```
  File "/firedrake/src/PyOP2/pyop2/parloop.py", line 421, in reduction_begin
    requests.append(self.comm.Iallreduce(glob._data, glob._buf, op=mpi_op))
AttributeError: 'Mat' object has no attribute '_data'
```

**Comment by JHopeCollins** [comment]
@ABaierReinio please can you try running with `"mat_type": "matfree"`?

Your solver parameters look very similar to the default parameters Firedrake sets for mixed spaces with Real space components, but the mat type is the obvious difference. You will need to use `AssembledPC` for the A00 block. See:

https://github.com/firedrakeproject/firedrake/blob/fefed5e1aee7805e39b22338e5d7229316e9f08b/firedrake/solving_utils.py#L81

@connorjward I think the problem might be that with `"mat_type": "nest"` Firedrake will default to using `sub_mat_type = "baij"` which might not be valid for a Real space:
https://github.com/firedrakeproject/firedrake/blob/fefed5e1aee7805e39b22338e5d7229316e9f08b/firedrake/parameters.py#L81

On that note, it looks like we don't support setting `sub_mat_type` from the PETSc options, which maybe we want to do. It isn't picked up here:
https://github.com/firedrakeproject/firedrake/blob/fefed5e1aee7805e39b22338e5d7229316e9f08b/firedrake/variational_solver.py#L225
Which means it isn't passed through here:
https://github.com/firedrakeproject/firedrake/blob/fefed5e1aee7805e39b22338e5d7229316e9f08b/firedrake/solving_utils.py#L489

If this does turn out to be the bug then we should also look at this line. The `mat_type` defaults to "nest" for Real blocks, and hence will have "baij" sub matrix types - I doubt anyone would pass `sub_mat_type` if they haven't passed `mat_type`. I don't think it will ever be hit for a solve because we set the matfree default whenever there is a real space, but it could be hit assembling matrices anywhere else.
https://github.com/firedrakeproject/firedrake/blob/fefed5e1aee7805e39b22338e5d7229316e9f08b/firedrake/assemble.py#L1264

**Comment by connorjward** [comment]
This is one of those annoying things where I have totally redone this bit in pyop3. The `mat_type` and `sub_mat_type` parameters are, in my opinion, poorly done.

In pyop3 I have it that `mat_type` is either a string (e.g. `"baij"`) or a list-of-lists (to signify a nested matrix). For example a nested matrix with Real blocks then has `mat_type`
```
[["aij", "dat"],
 ["dat", "global"]]
```
`sub_mat_type` isn't needed.

If there is a reasonable workaround I would be tempted to punt on this until pyop3 lands and we can make this a lot nicer then.

**Comment by JHopeCollins** [comment]
@ABaierReinio these parameters work for me, will this allow you to do what you need? I'm assuming your actual use case is more complex.

```
newton_solver_parameters = {                                                                  
    "snes_type" : "newtonls",                                                                 
    "snes_monitor" : "",                                                                      
    "snes_converged_reason" : "",                                                             
    "mat_type" : "matfree",                                                                   
    "ksp_type" : "fgmres",                                                                    
    "ksp_monitor" : "",                                                                       
    "pc_type": "fieldsplit",                                                                  
    "pc_fieldsplit_type": "schur",                                                            
    "pc_fieldsplit_schur_fact_type": "full",                                                  
    "pc_fieldsplit_0_fields": "0",                                                            
    "pc_fieldsplit_1_fields" : "1",                                                           
    "fieldsplit_0" : {                                                                        
    "pc_type": "python",                                                                      
        "pc_python_type": "firedrake.AssembledPC",                                            
        "assembled_ksp_type" : "preonly",                                                     
        "assembled_pc_type" : "lu",                                                           
        "assembled_pc_factor_mat_solver_type" : "mumps",                                      
    },                                                                                        
    "fieldsplit_1" : {                                                                        
        "ksp_type" : "gmres",                                                                 
        "max_it" : 1,                                                                         
        "convergence_test" : "skip"                                                           
    }                                                                                         
}
```

**Comment by connorjward** [comment]
> What are the "dat" and "global" strings in your pyop3 example? Can I pass them as petsc options?

The reason we have to use MATNEST for the Real space is because we use PyOP2 `Dats` and `Globals` for the (dense) rows and columns that correspond to it (explained in https://www.firedrakeproject.org/r-space.html#representing-matrices-involving-r). We therefore have special MATPYTHON types that wrap these objects ([eg](https://github.com/OP2/PyOP2/blob/31471a606a852aed250b05574d1fc2a2874eec31/pyop2/types/mat.py#L978)) and that's what `"dat"` and `"global"` represent here.

**Comment by JHopeCollins** [comment]
Oh, so each list in the list is the mat_types for the blocks on a single row?

**Comment by connorjward** [comment]
> Oh, so each list in the list is the mat_types for the blocks on a single row?

Yeah that's the idea. Happy to have feedback on it but I think it's clearer than the current approach. I don't like the lists of lists approach very much. It could also be a dict:
```py
{
  (0, 0): "aij",
  (0, 1): "dat",
  (1, 0): "dat",
  (1, 1): "global",
}
```

**Comment by JHopeCollins** [comment]
Whatever it is, there needs to be a way to pass it through the PETSc options for when you have deeply nested solvers.

**Comment by ABaierReinio** [comment]
> @ABaierReinio these parameters work for me, will this allow you to do what you need? I'm assuming your actual use case is more complex.
> 
> ```
> newton_solver_parameters = {                                                                  
>     "snes_type" : "newtonls",                                                                 
>     "snes_monitor" : "",                                                                      
>     "snes_converged_reason" : "",                                                             
>     "mat_type" : "matfree",                                                                   
>     "ksp_type" : "fgmres",                                                                    
>     "ksp_monitor" : "",                                                                       
>     "pc_type": "fieldsplit",                                                                  
>     "pc_fieldsplit_type": "schur",                                                            
>     "pc_fieldsplit_schur_fact_type": "full",                                                  
>     "pc_fieldsplit_0_fields": "0",                                                            
>     "pc_fieldsplit_1_fields" : "1",                                                           
>     "fieldsplit_0" : {                                                                        
>     "pc_type": "python",                                                                      
>         "pc_python_type": "firedrake.AssembledPC",                                            
>         "assembled_ksp_type" : "preonly",                                                     
>         "assembled_pc_type" : "lu",                                                           
>         "assembled_pc_factor_mat_solver_type" : "mumps",                                      
>     },                                                                                        
>     "fieldsplit_1" : {                                                                        
>         "ksp_type" : "gmres",                                                                 
>         "max_it" : 1,                                                                         
>         "convergence_test" : "skip"                                                           
>     }                                                                                         
> }
> ```

Thank you Josh, these parameters do seem to resolve the issue for the minimal working example I posted here. I guess this is more a workaround than an actual fix to the bug? I don't know enough of the technicalities behind all of this, so I'm happy to go with whatever works.

Unfortunately, in my actual more complicated use case, I'm using EquationBC. There currently appears to be another unrelated bug (which I'm currently discussing with @ksagiyam in Slack) that causes EquationBC to break when used with AssembledPC. So I won't be able to use this workaround until that other issue is also resolved.

In the meantime I've been using a different workaround. As I posted above, when using `"mat_type" : "nest"` the solver throws an error, but only after successfully doing one Newton iteration. So I've just wrapped the Newton solve in a try-catch block, and wrapped that in a while loop (it was a bit more complicated than that for reasons not worth mentioning here). Everything works when I do this! But obviously this is not a very satisfying way to proceed.

**Comment by JHopeCollins** [comment]
Ok, we can revisit the Real space problem once the EquationBC issue is worked through.

Depending on the exact point in the solve the error gets thrown, it might be nicer to set the snes iterations to 1, and skip any convergence test, then write the while loop to manually calculate the residual yourself rather than relying on the error.
snes_type ksponly might also work here, and avoids some code paths, depending on how the rhs gets set.

**Comment by ABaierReinio** [comment]
Thanks -- I wasn't able to get it working with `ksponly` but using `"snes_stol" : 1e100` I can ensure that it does at most one iteration, allowing me to avoid the try-except block. Until the bug is fixed I don't think it will get any better than this.

**Comment by connorjward** [comment]
@ABaierReinio I believe that I have fixed this in #3846.

**Comment by ABaierReinio** [comment]
Thanks Connor! With your fix I can now use `"mat_type" : "nest"` without any `AssembledPC` in the fieldsplit.

Funnily enough, Koki just fixed the issue I was having with `AssembledPC` (https://github.com/firedrakeproject/firedrake/pull/3842). So I now have the choice of using `"mat_type" : "nest"` without `AssembledPC` or `"mat_type" : "matfree"` with `AssembledPC` as Josh suggested. Is there any reason to use one over the other? Numerically they both give me the same answer.
